### PR TITLE
Add html object type for cards.

### PR DESCRIPTION
**Pull request by flackr**
Some cards contain mixtures of styling, tables, symbols, and other complex content. Given HTML is already capable of representing rich layout containing mixtures of these, it would be nice to be able to have a deck with portions of the content defined by HTML.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1407/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1407-pr-test`.

**Comment by flackr** [comment]
Not sure if this is already supported or not, but it would be great to be able to do variable replacement within the HTML as well so that you could template out a card layout with HTML substituting in parts.

**Comment by flackr** [comment]
So what I've done for security is load the content into a sandboxed iframe with no permissions added, see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox for details on the effect of this attribute. The frame has `pointer-events: none` applied to it so that events over the card face are still sent to the card widget.

Properties are currently applied currently by a regular expression substitution of `${name}` to the name property on the card or empty string if missing. Open to alternative ideas. We could also consider having a standard template for the frame (e.g. no need for every frame to specify `html, body { height: 100%; }` and we could even consider passing in variables, css styles, etc to a constructed document.

You can see in the example that assets can be included allowing for things like icons within blocks of text which is common on more complex game card rules.

**Comment by ArnoldSmith86** [comment]
It feels weird to me that you use `${PROPERTY x}` syntax and then still have to "import" them using `dynamicProperties`. Somehow logical in the context of card face objects but not what I would expect.

What do others think about it?

**Comment by flackr** [comment]
I figured out the difference in rendering is different default fonts. I need to figure out the right way to include virtualtabletop's css, specifically https://github.com/ArnoldSmith86/virtualtabletop/blob/main/client/css/fonts.css , from the embedded document and the rendering should be consistent.

File: client/js/widgets/card.js [hunk 1]
```diff
@@ -139,12 +139,33 @@ class Card extends Widget {
               objectDiv.style.backgroundImage = mapAssetURLs(`url("${object.value}")`);
             }
             objectDiv.style.backgroundColor = object.color || 'white';
+          } else if (object.type == 'html') {
+            // Prevent input from going to frame.
+            objectDiv.style.pointerEvents = 'none';
+            objectDiv.setAttribute('sandbox', 'allow-same-origin');
```
Comment (flackr) [inline client/js/widgets/card.js:145]: This `'allow-same-origin'` permission means that the sub-page can fetch resources as same-origin requests. We could probably avoid it by serving the fonts css and font files with CORS headers so that the frame can still be treated as cross origin. I'll look into adding this to the PR.

Comment (flackr) [reply client/js/widgets/card.js]: Done, I restored the full sandboxing and added CORS headers to the fonts.css and fonts so the cross origin frame can still fetch them.

**Comment by flackr** [comment]
I added `enlarge: 3` to test that as well, and discovered when showing the cards (especially the enlarged view) on Firefox there is an initial paint with the default fonts before the web fonts load and it re-renders. It's not the worst thing ever, but there are ways we could mitigate this:
1. Pass `allow-same-origin` to the [sandbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-sandbox) attribute. This would also allow removing the CORS headers added in server.mjs to the fonts. As a same-origin resource it's allowed to use the cached fonts for that origin. This seems ideal, as long as there are no dangerous URLs a malicious game creator could fetch from.
2. Use [font-display: block;](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display) in `fonts.css` to give the nested page longer to load used fonts. The downside is this may slow the initial render of the enlarged view.
3. Live with the switch of fonts. In most cases it won't be a significant change.
4. Inline the css / fonts. This would allow removing the CORS headers as it wouldn't need to fetch any external resources but is likely wasteful to inline all of the possible fonts that could be fetched.

**Comment by ArnoldSmith86** [comment]
**In theory**, `enlarge` could enlarge the actual widget, render a screenshot, shrink it again and display the screenshot instead of the current clone.

That wouldn't work with any animation though. Although it's weird anyway that animations aren't in sync with the current approach. It will also break `css` tricks that affect the enlarged widget but not the actual one (though they might be fixable by temporarily applying the `#enlarged` ID or whatever it's called.

**Comment by flackr** [comment]
Another option would be to load untrusted cards from a different domain and postMessage the html content to run to an empty frame running something like the following:

E.g. an `<iframe src="https://sandbox.virtualtabletop.io/card-frame.html">` would load the following:
```html
<html>
  <head>
    <style> /* Standard fonts and styles */ </style>
    <script>
window.addEventListener('message', (evt) => {
  document.body.innerHTML = evt.data;
}, {once: true});
    </script>
  </head>
  <body></body>
</html>
```

Then post a message with the contents of the card to the frame. It would have the standard cross origin frame restrictions.

This does complicate the serving of virtualtabletop since it would need to use an alternate domain for the secure card frame but we could always fall back to a same origin frame (or use tricks like wpt uses switching localhost with 127.0.0.1 - [link](https://github.com/web-platform-tests/wpt/blob/0c65de806d097cd459231f09c7d845ec5d5edfde/common/get-host-info.sub.js#L20)) for non-deployment testing scenarios.

**Comment by ArnoldSmith86** [comment]
> This seems ideal, as long as there are no dangerous URLs a malicious game creator could fetch from.

Is there more risk than the current way of just accepting any URL for images (for example in `css`)?

**Comment by flackr** [comment]
> > This seems ideal, as long as there are no dangerous URLs a malicious game creator could fetch from.
> 
> Is there more risk than the current way of just accepting any URL for images (for example in `css`)?

That's a very good point. It's already possible to craft same-origin credentialed requests through css so allowing it from the html frame doesn't introduce any new surfaces - especially without script.

**Comment by flackr** [comment]
> It feels weird to me that you use `${PROPERTY x}` syntax and then still have to "import" them using `dynamicProperties`. Somehow logical in the context of card face objects but not what I would expect.
> 
> What do others think about it?

It can also refer to properties on the object itself. Not sure if that use case serves much value, but it follows the pattern of how `dynamicProperties` defines which object properties are overridden by which dynamic properties. I think it would be nice not to have to say `${PROPERTY` either in the `html` or `css` properties but I feel like they should be consistent with each other.

So I think the proposal would be that `${x}` in either `css` or `html` would refer to that object's property value - whether that came from the object itself or the dynamic property from the `cardType`? This sounds good to me.

**Comment by ArnoldSmith86** [comment]
I would probably only support `${PROPERTY x}` and make that pull from the widget. Because that is what `css` does.

**Comment by flackr** [comment]
> I would probably only support `${PROPERTY x}` and make that pull from the widget. Because that is what `css` does.

Sorry, to be clear, if `${PROPERTY x}` pulls from the widget, how do you pull from the dynamic properties?

**Comment by ArnoldSmith86** [comment]
The dynamic properties are widget properties anyway. And static properties are static so you could just use their value in the HTML.

**Comment by flackr** [comment]
> The dynamic properties are widget properties anyway. And static properties are static so you could just use their value in the HTML.

When I was implementing this, the way I understood the code was that the dynamic properties are not widget properties. They are applied to the `object` within the card face.

**Comment by ArnoldSmith86** [comment]
Yes. _From_ the widget.

**Comment by flackr** [comment]
> Yes. _From_ the widget.

Dynamic properties also come from the card type though .. which isn't on the widget is it?

**Comment by ArnoldSmith86** [comment]
It's all the same. cardDefaults and cardType properties override widget properties. They are all returned by a GET.

**Comment by flackr** [comment]
> It's all the same. cardDefaults and cardType properties override widget properties. They are all returned by a GET.
> 
> So if you use PROPERTY syntax in `css`, you can also access cardType and cardDefaults.

Thanks for the explanation. I think when I was working on this I hadn't quite figured that out from the code. This shoudl simplify it a bit :-).

**Comment by flackr** [comment]
@ArnoldSmith86 based on our discussion I made `${PROPERTY x}` pull `x` off of the widget, and removed the custom logic for css properties, so now this should all be consistent. It also no longer requires the properties to be in dynamicProperties because the object properties aren't the things being replaced.

Doing all this I just rediscovered part of what lead me down this rabbit-hole in the first place, which is that dynamic properties trigger a call to setValue which updates the card face but we don't track changes to any other used properties. I think this is a pre-existing bug that you could see via css property replacement on objects which is now also exposed by properties referenced from the html value as well.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -441,15 +442,15 @@ export class Widget extends StateManaged {
...
     for(const match of String(css).matchAll(/\$\{PROPERTY ([A-Za-z0-9_-]+)\}/g)) {
       css = css.replace(match[0], this.get(match[1]));
-      this.propertiesUsedInCSS.push(match[1]);
+      usedProperties.add(match[1]);
     }
     return css;
   }
 
-  cssToStylesheet(css, nested = false) {
+  cssToStylesheet(css, usedProperties, nested = false) {
```
Comment (flackr) [inline client/js/widgets/widget.js:453]: @ArnoldSmith86 I was torn between adding a parameter to track used properties or changing the return value to be an object. I think the latter may be slightly cleaner from an API perspective but this is technically slightly more efficient (don't have to create and merge sets) and a bit of a smaller code change. Wanted to get your thoughts. WDYT?

I.e. the two options are:
1. As is in this PR, usedProperties is a Set pased in that the css methods add to when a property is referenced.
2. Or the css methods return a structure such as { css: 'original css text return value', used: <set of used properties> }

**Comment by 96LawDawg** [comment]
Comments:

1) The video still does not seem to work for Chromium browsers.  Works fine in Firefox.
2) This does not play nicely with enlarge.  May not be fixable.  May not be a problem.  Just pointing that out.
3) Please suggested some wiki edits to document this.  I don't understand it well enough to try.  Understanding it better will also let me test it better.

**Comment by ArnoldSmith86** [comment]
> This does not play nicely with enlarge. May not be fixable. May not be a problem. Just pointing that out.

What happens?

**Comment by 96LawDawg** [comment]
The enlarge view goes blank while you are dragging it, but the outline of the card enlarge area stays there.

**Comment by flackr** [comment]
Thanks so much for the initial testing comments @96LawDawg. Some initial comments but will follow up once I have time to debug.

> 1. The video still does not seem to work for Chromium browsers.  Works fine in Firefox.

If people are fine with it, I'll probably not worry about video yet in the initial PR. There's a lot of value in the non-video capabilities.

> 2. This does not play nicely with enlarge.  May not be fixable.  May not be a problem.  Just pointing that out.

I think right now while dragging it is reconstructing the card face object every frame. I should be able to optimize this which will fix this as well.

> 3. Please suggested some wiki edits to document this.  I don't understand it well enough to try.  Understanding it better will also let me test it better.

Will do!

**Comment by flackr** [comment]
@96LawDawg My latest commit avoids recreating the drag element dom if the delta doesn't include any properties which would lead to a visual change of the enlarged view. This prevents the flicker / disappearing during dragging that you were seeing.

Documentation would probably go in https://github.com/ArnoldSmith86/virtualtabletop/wiki/Cards-and-Decks
We could add a third object to the example with an image and text object, as follows (ellipses used to ellide existing content):

```
  "faceTemplates": [
    {
      ...
    },
    {
      "objects": [
       {
          "type": "image",
          ...
        },
        {
          "type": "text",
          ...
        },
        {
          "type": "html",
          "css": {
            "body": { "display": "flex"; "flex-direction": "column" },
            ".desc": { "flex": 1 },
            ".footer": { "font-size": "12px" },
          }
          "x": 0,
          "y": 60,
          "width": 140,
          "height": 64,
          "value": "<div class=desc>${PROPERTY description}</div><div class=footer>${PROPERTY summary}</div>"
        },
      ]
    }
  ]
```

...

The third element on the face is html. This allows for styled text, embedded images, and generally anything that is allowed in html. This is positioned at the middle of the card vertically flowing towards the end of the card. The particular description and summary is given by the `description` and `summary` properties of the card.

This is really a brief summary, but there are many ways this can be used to build flowing content for cards, e.g. the `description` property can contain embedded images: `"description": "This provides 4 ."`, tables, etc. There are examples of some ideas in the PR test.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.

One request, can you demo how to use contentEditable on a card, or is that not possible?

**Comment by flackr** [comment]
> Approved through test.
> 
> One request, can you demo how to use contentEditable on a card, or is that not possible?

This isn't possible right now. To make sure that things are still draggable I set `pointer-events: none` on the frame, becuase if input events are sent into the frame they don't bubble out (and don't start drags). It might be doable with some script coordination but this also disables script execution in the frame for extra security.

TLDR; This is something we could explore in the future but not possible yet.

**Comment by 96LawDawg** [comment]
Revised Decks tutorial (@flackr, please add to your PR; I still don't see your branches).

Modifies the original variant 0 to be basic and adds a new variant 1 for HTML.  I just copied and pasted the stuff from @flackr's demo room as I still haven't looked at it enough to make a more comprehensive example.

[Decks.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/10356550/Decks.zip)

**Review by RaphaelAlvez — APPROVED** [review]
Approved in meeting
just minor change requests about tutorials

**Comment by 96LawDawg** [comment]
As discussed in meeting, here is the file for tutorial changes we discussed.  Please use this file in place of the existing Decks tutorial.  Also, delete the faceTemplate tutorial as it has been merged into this one.  After those 2 changes, this is ready to be merged.

[Decks.zip](https://github.com/ArnoldSmith86/virtualtabletop/files/10416134/Decks.zip)

**Comment by flackr** [comment]
Thanks @96LawDawg, changes made.
